spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy

```diff
@@ -0,0 +1,1237 @@
+package spock.util
+
+import org.spockframework.compat.groovy2.GroovyCodeVisitorCompat
+
+import java.lang.reflect.Modifier
+import java.security.CodeSource
+
+import groovy.transform.CompileStatic
+import groovy.transform.Immutable
+import groovy.transform.TupleConstructor
+import org.apache.groovy.io.StringBuilderWriter
+import org.codehaus.groovy.ast.*
+import org.codehaus.groovy.ast.expr.*
+import org.codehaus.groovy.ast.stmt.*
+import org.codehaus.groovy.classgen.*
+import org.codehaus.groovy.control.*
+import org.codehaus.groovy.syntax.Types
+
+/*
+ * This has been adapted from https://github.com/apache/groovy/blob/5d2944523f198d96b6515e85a24d2b4e43ce665f/subprojects/groovy-console/src/main/groovy/groovy/console/ui/AstNodeToScriptAdapter.groovy
+ * and made backwards compatible with groovy 2.5 via GroovyCodeVisitorCompat.
+ *
+ * - Replaced getter access by property access
+ * - Removed trailing whitespaces when directly followed by printLineBreak()
+ * - Special handling of AnnotationConstantExpression in visitConstantExpression
+ */
+
+/**
+ * This class takes Groovy source code, compiles it to a specific compile phase, and then decompiles it
+ * back to the groovy source.
+ */
+@CompileStatic
+class SourceToAstNodeAndSourceConverter {
+
+  /**
+   * This method takes source code, compiles it, then reverses it back to source.
+   *
+   * @param script
+   *    the source code to be compiled. If invalid, a compile error occurs
+   * @param compilePhase
+   *    the CompilePhase. Must be an int mapped in {@link CompilePhase}
+   * @param showSet
+   *    the set of code to include in the rendering
+   * @param classLoader
+   *    (optional) the classloader to use. If missing/null then the current is used.
+   *    This parameter enables things like ASTBrowser to invoke this with the correct classpath
+   * @param config
+   *    optional compiler configuration
+   * @returns the source code from the AST state and the captured ast nodes
+   */
+
+  AstResult compileScript(String script, int compilePhase, Set<Show> showSet, ClassLoader classLoader = null, CompilerConfiguration config = null) {
+
+    def writer = new StringBuilderWriter()
```

Comment:
It wouldn't support rendering package and imports right? 
Any other things that wouldn't be supported?

We can decide, that we are not interested in those things anyway.
The final thing would need to support multiple classes., not just one.

Reply:
I've compared the two compilation methods and they are doing almost the same.
One difference is that you cannot give in an own class loader, but it uses a `new GroovyClassLoader()` but as far as I understood the own class loader in the embedded spec compiler is just to get `Class` instances out easily and you simply reused it because it was there anyway.
The other is that you don't get package statement and imports.

As the remaining things are identical I think there should be no further differences and I would agree that we don't necessarily need the package and imports ever in those tests.
And if we want them, we should anyway change the logic a bit, because currently the package and imports are repeated for every class node.

Reply:
For the time being I would keep the current variant